### DISCUSSION TITLE
Merk40T on 5 inch display.

### DISCUSSION THREAD

**Discussion by edcasati** [meerk40t/meerk40t#1944] [Ideas]
Question: Is there an alternate interface that will run comfortably on a 5 inch screen?

Question explained: 
I just received my first K40, and haven't even set it up yet. 
I plan on running  Meerk40T on a Raspberry PI 3 . The one that I will be using already has a 5 inch touchscreen attached. Trying to run the Merk40T interface on a 5 inch screen is ludicrous, so I set it up to connect via VNC to a regular computer.
It would if great if there was an alternate display mode that fits on a 5 inch screen, with just the essentials, like jogging, homing, pick a file to laser, test fire.  (Long term maybe it could even be expanded using the GPIO pins on the Pi to show water temps, etc)
There are lots of examples of these simplified screen used on 3D printers, so you get the idea.

Edward

**Comment by tatarize** [comment]
No. You could shut down the ribbonbar. @joerlane also tries to use something like that, and equally has very little display to see. I'm not sure what such an alternative display would use. The ribbonbar is a bit essential but I'm not really sure what this alternative display would use. I would assume we would generally not need the ribbonbar tools. But, we'd want to basically load a job and maybe do the basic work of executing it. 

The GPIO stuff would likely be able to be hooked up with a plugin since there's a rather notable plugin architecture. Though apparently there's a lot of different water temp signals that you could have, so it wouldn't really be able to standardize that very obviously. I assume the laser-panel pane would be close to this sort of idealized 5in display view. You can start jobs, preview them (probably also pretty bad), jog, move, (no fire laser pulse), but it wouldn't be too terrible. 

I assume you'd want basically a just the center pane. Put into a notebook, with a series of pages. So you'd have a `load` page, a scene page, an execute page. a jog page. And setup a bit like that? Which would be like scene + laser_pane in the same notebook in the center of the screen without the ribbon-bar etc.

**Reply by tatarize** [reply]
I assume we'd just want a bunch of rehashed obvious in a little notebook as the main-window probably triggered with a CLI command to use that rather than the default main window. I assume this wouldn't really even require AUI and we'd probably be better off without it. So you'd just get a bunch of common panes/panels.
So a bit like laser-panel with scene and open-file.

Perhaps running any loaded file up to the point of execution and letting you have a simulate tab, perhaps instead of scene. It sort of depends on the workflow I'd think. Camera would also be sort of nice there. 

<img width="276" alt="scene" src="https://github.com/meerk40t/meerk40t/assets/3302478/4ff0312d-8ba0-4c6b-81ca-772089d7a629">

Probably:
Laser-Device, Load.  Simulate. Jog, Move.  Camera. Each with their own page. Though simulate might need to be simplified down.

**Comment by edcasati** [comment]
Are there keyboard shortcuts for moving the head, sending home, test firing laser, etc.? If so, creating a 'pendant' like interface could be possible doable? Most pendants just send keystrokes through pushbuttons.
Just thinking out loud here..... I am amazed and feel humbled by the system that you have created already!

________________________________

**Comment by Sophist-UK** [comment]
I am personally unclear whether we would want a full MK experience squished into a 5" display, or instead a rPi version of a remote client with an entirely new GUI focused on setting origin / doing hull shape outlines etc. i.e. all the stuff you do when you are preparing the actual burn.

**Comment by tatarize** [comment]
Not a full experience. You likely can't (or shouldn't) do that.. You'd get a highly limited experience. The referenced prior art is Octoprint TouchUI or so which can have a little limited touch screen capability. You'd want something which would work a lot more like a very advanced looking ruida set of controls, so more like if you had a really great little UI to control your laser. So it would appear much more like an embedded device UI than the current MeerK40t UI.

![touchui](https://github.com/meerk40t/meerk40t/assets/3302478/5ba50890-cabe-4153-93d0-ecb17937eac9)

You'd basically *just* focus on controlling the device, anything that is basically concerned with the design would not be part of the display. We'd likely just launch a single window and give it just the GUI parts we want.

And each page of the notebook would have limited control aspects.

* Load design
* Jog, Adjust design position, outline
* View/Simulate the design
* Execute
* Camera

There are certainly some aspects of this which would *clearly* work better as a plugin. For example, I certainly can't/won't include Flask as part of the full program because it has a lot of dependencies. But, there's obviously some clear advantages to doing that and making the less obvious part of this UI also run a Flask UI. Especially file loading would be *really* helpful with that. So the load file would basically be load recent or these svgs (or other files) located in the current working directory.

https://github.com/tiger12506/restk40t

This old plugin code from a couple years ago, comes to mind. In theory, we could probably use that to more directly send the files we'd like to run on laser. And really done right that would make the 5in display more of a basic extension of that. Since really the biggest hamper on your workflow will be sending the file to the Raspberry pi (since you wouldn't design it there). Then the 5in display would be select a file, jog the position (using current position or absolute), and then fire off the laser.

But, at the most basic we'd have a window consisting of a set of notebook pages. To start we'd probably do load, jog, and start, and could probably do about the same thing with flask (though that would require being in a plugin). And we'd sort of want library stuff. It would be about as basic as K40 Whisperer, and heck we could largely give it the same sort of functionality since most of that functionality we just have laying around now. We also already in 0.9.x allow `camera1 start server 4445` or whatever which would launch the camera1, start the execution, then start serving the camera on `localhost:4445` or so. Generally this is so Raspberry Pis with cameras can be useful. But, for a sort of http/api through flask this would let you toggle the camera(s) on and view them embedded in a webpage.

Everything else gets easily control with the internal console and that stuff is pretty easy to hook together. You wouldn't really get a good view of the scene over the web-interface, but could on the local UI interface. 

In theory you could want slightly different UI, for slightly different configured devices. There's clearly going to be advantages to allowing lighting jobs for balor, that you wouldn't have for other lasers you'd want to control. Since hitting the button and being able to outline the lighting job. Or do tap to continue on the touch screen would also be quite helpful.


---

Codewise I think we'd just let you send a different command for a simplified display which would just be a single notebook window with dynamically created pages. And we'd give you bare minimum for the UI for that. We can add things like a navigation pane with a couple lines of code there. Open, Drag, Execute seem like they'd suffice as a minimum.

**Comment by Sophist-UK** [comment]
I think that the idea of using an existing embedded controller (from e.g. Ruida controllers) as functional inspiration is an excellent idea and IMO there would be a lot of benefit to having a common learning/curve / experience from having something that looks and feels quite similar.

**Reply by edcasati** [reply]
When I first started with this 5" screen idea, I was building my own laser system from scratch. I got pretty far on the 5 inch screen idea, actually getting things to move around, read files, etc., but then the project went by the wayside. I used Python and Tkinter  to send commands to the Octoprint console which ran in the background, File management was done  wirelessly to the Raspberry Pi. When I found Meerk40T and decided to use it, the project came back to my mind.
An interface like this is great f you want to run the system 'standalone' where you just load a file, position with a red dot frame, then execute an existing file, without the need of an external computer at all. 
Anyway, it's just an idea.

**Reply by Sophist-UK** [reply]
@tatarize had an idea ages ago to write a cut-down remote-only controller - the idea being to produce it as an Android app so that you can connect your K40 to your mobile phone USB and get Ruida-like control screen, but if David ever gets the time to produce this then undoubtedly the same code could run on a rPi with or without a touch screen.

**Reply by tatarize** [reply]
![image](https://github.com/meerk40t/meerk40t/assets/3302478/0fe80e7b-92c7-4320-a630-36fbba2b1b7b)

It's actually a bit hard to get everything there correctly and some of the gui stuff expects the main window to be the real main window but this sort of works.

**Comment by tatarize** [comment]
Okay, cursory attempt in #1948

Mostly hit a couple of these Frames that sort of expected that wxMain has loaded and registered some settings. But, these are bugs that are generally just impossible to detect other than a project like this. Frames/Windows are generally expected to be self-contained.

The only way to load the file is to type "load <filename>" in the console, at least currently.

The buttons for the current `LaserPanel` might be too small and complex.

![image](https://github.com/meerk40t/meerk40t/assets/3302478/55849df9-49db-44f8-b77c-83b50fb75f2d)
![image](https://github.com/meerk40t/meerk40t/assets/3302478/b4089922-2add-418a-8cf4-d6054b9311bb)
![image](https://github.com/meerk40t/meerk40t/assets/3302478/01c24070-cd16-40ca-bd58-126f108aea75)


There's a few things which aren't really right. Simulate for example launches the real simulate window. Which isn't bad but that's bigger than this little one off window. I tried to start it and it crashed from a minor missing setting that is used internally (laser_hold) but which is loaded by the Main class but not the Simple class. I also would need to port in some bits of code so it shutsdown on close etc.

But, in theory this would work.

---

The biggest problem I see here is that I am not good at UI, and this seems like it needs a bunch of UX stuff and maybe even some custom panels.  I can dump whatever panels exist into this UI and quickly fix any bugs here. We can even make the panels differ slightly from the window versions by adding in specific custom calls. But I feel like it needs big Fisher-Price toy buttons or some prior art to fairly strictly adhere to.

**Reply by tatarize** [reply]
In order to run this, switch to the branch `simple-ui` and execute the code with `meerk40t.py --simpleui` and it'll launch with this window rather than the regular main one. If there's a problem you need to quit and troubleshoot in the fuller window mode or whatever.

**Comment by edcasati** [comment]
I found it once before but I can't find it anymore.
What image collection is used for the 'jog' screen?

**Reply by tatarize** [reply]
We use icons8 icons usually in iOS Glyph or Material Filled forms, we use the free license. 

https://icons8.com/icons/ios-glyphs

**Reply by tatarize** [reply]
https://github.com/meerk40t/meerk40t/wiki/Tech:-Adding-Icons-to-a-MeerK40t-Module

**Reply by tatarize** [reply]
A lot of them at least the useful ones are in https://github.com/meerk40t/meerk40t/blob/main/meerk40t/gui/icons.py

Which basically stores all the different png icons in base64 python form for ease of asset management. The icons.py thing itself does some nice things to let the icons be heavily customized including tint, flip, rotation, size, etc.

**Reply by tatarize** [reply]
If you're considering doing a mockup of the different frames That'd totally work. Most wxWidgets UI design is done with wxGlade:

https://sourceforge.net/projects/wxglade/

The internal stuff is basically all just console calls and extremely modular so really, anything pretty could be easily made to work. Even ugly things could be made to work.

I'm not sure the file thing wouldn't need it's own file manager thing letting you check through different directories. I see that's what the TouchUI for the Octoprint does, and I don't know the screen limitations but @joerlane says he needed to go double and even triple pixels to get anything coherent.

If you just want smaller icons they can be made smaller with a simple call to the icon_size building the jog screen.

**Reply by tatarize** [reply]
![image](https://github.com/meerk40t/meerk40t/assets/3302478/fe38612e-97bf-4456-aeda-164bc576ae2e)

For example, here's the jog panel at 20px each. This is also done in the LaserPanel of the main version with the jog and move put together on a thing.

**Comment by tatarize** [comment]
![image](https://github.com/meerk40t/meerk40t/assets/3302478/23630670-8dc3-4096-973f-8f7b620c11a8)

Not sure if that's enough to actually get things to work, the load project loads the default file selection which I do not know if that's terrible on a 5in display or not. We could create dynamic buttons to do file selection etc.

We could probably do file viewing, though the scene itself seemed like it had a lot of settings that were registered in the main dialog rather than the scene and thus were a bit more hinky. Jog works fine. It might need somethings like a very basic generic controller to see what's going on.

But, that should actually be enough to load a job and start the job.

I could also force add a `place current` node which would (though you can't see the tree) cause the project to be based on the current position. So jogging around would actually affect where the data got placed.

---

If I were good at UI, I probably wouldn't have one with many badly hidden features as to not be overwhelming.

**Comment by Sophist-UK** [comment]
Designing a good UI when you are showing all the functionality is difficult enough IMO, but deciding which functionality to make visible in a cut down UI and selecting those in a way that those that are shown and those that aren't are both a consistent whole has got to be much harder (hence my suggestion to treat it as a way of giving a UI to effectively an enhanced controller (aka Ruida) as a remote driver rather than a full blown MK execution).

**Reply by tatarize** [reply]
I was thinking of maybe adopting the regular sort of K40 Whisperer workflow. While I don't love it for a full program since it's decidedly limited. Decidedly limited is what you'd need in a simple UI like this. 

* Load the file.
* Show a scene of the expected simulated result. 
* Allow the user to do some basic interactions with that simulation to see how it would execute.
* Allow the user to jog the laser around and use the laser position as a current placement.
* Let the user start the job.

The problems here appear to be:
* User has no control over the resulting work product. (this is acceptable)
* User has no control to allow for the adjusting of speed for the raster, engrave, or cut operations.
* User would need to visualize the cut-code meaning writing up a simple simulation scene (probably with time-based object selection).

You're basically hoping that classification is fairly perfect, while it's now pretty good, it's not perfect. To create a single coherent-ish simple-ui dialog the only thing I can really think of is Whisperer's sort of interface. And while it would be trivial to make a place you enter a speed and execute rasters, engraves, and cuts. That's somewhat questionable. But, it seems like the current panes are more slated towards the more complex workflows, or rather a fairly complete set of workflows.

But, I could absolutely duplicate the typical whisperer workflow. It wouldn't be too hard to write a version of simulation that would give you a seekbar, presentation of the project and let you jog the laser. I could even do that with buttons in the scene to make good use of the space. 

The bones in meerk40t are such that I could whip the whole thing up in a night. But, you'd think you'd want the ability to jog the laser into place. Move the project around. Adjust the speed and power. Simulate it very generally. Estimate the time to execute it. Execute it. Monitor the job. halt/pause the execution. Maybe access the camera(s).

**Comment by edcasati** [comment]
I'm just messing around trying to pass keypresses to the console. Don't worry, I wont ask for any changes at all to the program.
If I come up with anything, I'll share a video of it in action.

________________________________

**Reply by tatarize** [reply]
I'd prefer you have an entirely thought-out interface with the exact things that make the cut and the things that don't that would maximally allow a user to load, visualize, execute, and monitor a job (within the extremely limited space). Like here's the mockup of exactly what needs to happen and which UI aspects to use. 

Changes to the program are easy. There's a reason why there's secret features to do things like draw animated ribbons and old-pendulum based Victorian era drawing machines. And why pr #1948 is 245 lines of code of which 232 lines is the simpleui window. Most of my code has focused on making that stuff easy to do. 

I have no trouble making the code show or do anything, on any lasers etc. The issue is really designing the UI for the what makes the cut. There's the old saying "I wrote you a long letter. If I had more time, I would have written a shorter letter." Good UI design is really hard, and I could probably hammer nearly anything in very short order. The code is designed to be GUI-agnostic and let you run everything you need from the console. There's a pretty worldclass visual api widget/scene system etc. --- If you could mock-up/detailed-explain what you want I could probably code it up fairly short order. That's really what's missing from this.

**Reply by tiger12506** [reply]
Tat, isn't this one of the reasons you developed the scene-based tool framework?

**Reply by tatarize** [reply]
Somewhat. I'm not opposed to just coding up a new scene. In theory all the parts could be done within a single scene window, position, load file, preview, execute. Then some stop and pause buttons while it's running. And maybe a simple method to set power and speed (I don't really see how). It also might be nice if the resulting program could work for normal users who just want an extremely simple UI anyway. I figure we'd do away with all the regular scene design elements and just have a selected set of elements in the scene. This could even be done at the simulation or using the same scene for both the simulated and pre-simulated view of the elements (though really we could probably just update the speeds within the cutcode and show that). 

Building the interface stuff is the easy part, designing that sort of interface would be the hard-part. But, in theory if `simulation` had a bit of an ability to set speeds/power and move the location around a little bit, and maybe load the initial design (and didn't have that giant space suck at the bottom there) it would be somewhat close.

**Comment by edcasati** [comment]
I've been playing with the concept, and I have a rudimentary 'move' screen running on the touchscreen. But now I'm kinda stuck, because I have no idea how to pipe information between two sessions. I have 'meerk40t -c' (console mode) on one session and it has the two '>>'  where you would type a command. and I can print commands from my LCD screen to the session where my routine is running.... but how do I get both of them talk to each other.
Thinking back to when I created something like this for my other laser project, I now remember that I could do it because Octoprint supports REST, so I was making HTTP posts to the Octoprint server. [](https://docs.octoprint.org/en/master/api/index.html) 
![move_screen](https://github.com/meerk40t/meerk40t/assets/9693192/895fcd28-1dd1-4dd1-8bc3-35ffc334d502)

**Reply by tatarize** [reply]
There's a rather extensive API but it's not a REST API. To be more rest-like it would need flask. You would only really need a plugin that would relay the commands to the console since you can do everything through the console there. You could also start up a consoleserver. `-e consoleserver` or `-e "consoleserver -p <port>"` would let you start a consoleserver on the given port you could then connect to that and send whatever commands you want over TCP and it would work like a console.

**Reply by tatarize** [reply]
Likewise this code is a bit old but gives an basic plugin setup.

https://github.com/tiger12506/restk40t/blob/master/restk40t/main.py

A less on the money, more modern example is given:

https://github.com/meerk40t/meerk40t-example-extension

Basically you can write a quick python script with an entry point and it will be loaded with meerk40t when meerk40t starts. This includes writing things like a touchscreen that could control the device, or a REMI interface ( https://github.com/rawpython/remi ) or flask interface.

**Reply by tatarize** [reply]
```python
from http.server import BaseHTTPRequestHandler, HTTPServer


def plugin(kernel, lifecycle):
    """
    This is our main plugin. It provides examples of every lifecycle event and what they do and are used for. Many of
    these events are simply to make sure some module events occur after or before other module events. The lifecycles
    also permit listeners to attach and detach during the lifecycle of a module, and insures everything can interact
    smoothly.

    :param kernel:
    :param lifecycle:
    :return:
    """
    if lifecycle == 'register':
        """
        Register our various processes. These should modify the registered values within meerk40t. This stage is
        used for general purpose lookup registrations.
        """

        class PostHandler(BaseHTTPRequestHandler):
            def do_POST(self):
                content_length = int(self.headers['Content-Length'])
                post_data = self.rfile.read(content_length).decode('utf-8')
                print(post_data)
                kernel.console(f"{post_data}\n")
                self.send_response(200)
                self.send_header('Content-type', 'text/plain')
                self.end_headers()
                self.wfile.write(b'Received POST request')

        def start_server():
            host = 'localhost'
            port = 8000
            server = HTTPServer((host, port), PostHandler)
            print(f'Starting server at {host}:{port}')
            server.serve_forever()

        kernel.threaded(start_server, daemon=True)
```

Here, for example is a rest-api of post commands.
`wget --post-data "scene rotate 44deg" http://localhost:8000`

Would, for example rotate your scene. It just needs a pip install on the same machine that is running meerk40t (if running from source).

[meerk40t-rest.zip](https://github.com/meerk40t/meerk40t/files/11681877/meerk40t-rest.zip)

**Reply by tatarize** [reply]
Likewise the commands for `right 5mm` `down 5mm` etc. would control your laser per-normal. Though there's probably a lot more obvious ways to hook this up when you see that it just gives your plugin access to the kernel and it can do basically anything meerk40t can do.

**Comment by edcasati** [comment]
what is a lhyserver? what is grblmock?

**Reply by tiger12506** [reply]
lhyserver is meek40t-speak for a TCP/IP server that listens for k40 code (such as presented by another instance of meerk40t on another machine) and passes it onto a physically connected laser. Very useful if, for example, you want to have a raspberry pi directly connected to your laser, and want to connect to it over the network to control the laser.

grblmock is (likely) a mode that "mock"s a real grbl device without actually needing areal grbl device connected. The goal being to test meerk40t is working properly without actual access to a real grbl laser.

**Reply by tiger12506** [reply]
You'd probably be better served by consoleserver, as Tat recommended. You could setup meerk40t to run consoleserver on port whatever, then in the backend for your GUI you made, you would just make a normal TCP/IP connection to that ip and port combination, and send console commands to meerk40t in plain text.
lhyserver is meant for direct sending of k40 code between two instances of meerk40t.

**Comment by edcasati** [comment]
Thank you for the info. I just saw them listed in the manual and was wondering if they could be of any use.

I am not a programmer by trade, and even barely a hobbyist, so much of what you say goes waaaay over my head and it takes me a while to somewhat understand. I really appreciate the effort that you are all making to help me in this project of mine.

(Retired old man, disabled and really looking for something to keep my brain alive and occupied, but it is a challenge since I can only use a keyboard 'sometimes' and a mouse even less often)

Thanks again!

**Reply by tiger12506** [reply]
How did you make this? If it's possible to share it, then I could probably give you instructions on how to hook it up to meerk40t. (i.e. I'd give you an example of how to make one button work, which you could use to make the rest of the buttons work similarly.)
<img src=https://github.com/meerk40t/meerk40t/assets/1036922/f46c4564-0d02-40e4-9bc9-52fd9287e584 width=40% />

**Comment by edcasati** [comment]
here is some basic functionality, still has a long way to go.
You click (touch) a button it sends the appropriate instruction to the console.
The numbered buttons set how far to move with each click (they need to become radio buttons)
What I have no clue on how to do is to print to the Meerk40T console so it accepts the instructions.

https://youtu.be/NS48YIL6ZoI

**Comment by edcasati** [comment]
Great news! I got it working... somewhat. Unstable as hell, but I can click a button on the screen and it echoes on the consoleserver. Works once or twice then crashes. But it's progress. Now to connect the display to the K40 and see if it moves.

**Reply by tatarize** [reply]
Bravo. Though that seems a bit convoluted to get there. Seems like running flask or something on the device running the laser could be a bit less jerryrigged.  Is the crash on MeerK40t's side, if so I could fix it? I didn't specifically check the use case of quickly connecting to console server sending some commands and then instantly quitting without even reading the response messages. --- So it could have some meerk40t crash in there.

**Reply by edcasati** [reply]
I'll investigate some more. It has been a fun adventure.
It basically ended up being 10 lines of code at the beginning of the program to redirect the 'print statement.

ChatGPT gave me the code.
```python
#--------------------------------
import sys
import telnetlib

class TelnetPrint:
    def __init__(self, host, port):
        self.tn = telnetlib.Telnet(host, port)

    def write(self, message):
        self.tn.write(message.encode('utf-8'))

    def flush(self):
        pass

# Redirecting print statements to Telnet
def redirect_print_to_telnet(host, port):
    sys.stdout = TelnetPrint(host, port)
#------------------------------------

# Usage example
redirect_print_to_telnet('localhost', 23)
print("Hello, World!")
```
Right now it doesn't close the connection after the send.

One problem that I have is getting the server to start listening reliably. Maybe I am not starting it correctly?

-------------------------------------
edcas@meerk40t:~ $ sudo meerk40t -c
wxMeerK40t plugin could not load because wxPython is not installed.
DXF plugin could not load because ezdxf is not installed.
OpenCV is not installed. Disabling Camera. Install with:
        pip install opencv-python-headless
>>consoleserver
[19:57:02]     MeerK40t 0.8.4000 pkg console server on port: 23
>>[19:57:02]     Could not start listening.
^Z
[3]+  Stopped                 sudo meerk40t -c
edcas@meerk40t:~ $

**Reply by tatarize** [reply]
Hm. Tried it a bit. Ran several times seemed to work fine.

**Reply by tatarize** [reply]
```
[18:14:20] Hello, World!
[18:14:20]     hello, is not a registered command in this context: Base
[18:14:20]     Connection to ('127.0.0.1', 49300) was closed.
[18:15:04]     Socket Connected: ('127.0.0.1', 49331)
[18:15:04] Hello, World!
[18:15:04]     Listening console-server on port 23...
[18:15:04]     hello, is not a registered command in this context: Base
[18:15:04]     Connection to ('127.0.0.1', 49331) was closed.
```

I surrounded your code with \`\`\`python \`\`\` (easier to read).

**Reply by tatarize** [reply]
You should be fine with `meerk40t -Ze consoleserver` I would think. Does it say anything if it doesn't start up correctly?

**Reply by edcasati** [reply]
You are correct in your first assumption.  As long as I click slowly things seem stable. Is there any kind of return code generated when a motion is finished?

**Reply by tatarize** [reply]
It might be the command. You likely are disconnecting very quickly and not seeing the reply. You could try:
`move_relative -f 1mm 0` etc. As a rule the jogging commands do not execute if there's anything in the spooling. The -f is short for `--force`

```
[18:20:54]     	move_relative <dx:Length> <dy:Length>
[18:20:54]     	(spooler) -> move_relative -> (spooler)
[18:20:54]     	Argument: Length 'dx':
[18:20:54]     		change in x
[18:20:54]     	Argument: Length 'dy':
[18:20:54]     		change in y
[18:20:54]     	Option: bool ('--force', '-f'):
```

**Reply by tatarize** [reply]
Also, there's a chance the routine isn't actually forcing the required `\n` at the end of the lines. Since telnet and other things sent through tcp could send partial commands it will actually buffer that. And the processing of the text won't care until it gets to `\n` whereas when you call `context("<command>")` it may not matter. But, through console server it might need an explicit end of line. I'll doublecheck I guess since I have your script and a laser. (And my current work is kind of annoying me).

**Reply by tatarize** [reply]
With your code as is, I'm getting fairly consistent results. Though I'm not hitting the button, but it seems like you'll have some lag to telnet in there a bit.

**Reply by edcasati** [reply]
I have to stop for now. Thank you for all the help.

**Reply by edcasati** [reply]
I'm making great progress on this. Update soon.

**Reply by tatarize** [reply]
ok

**Comment by tiger12506** [comment]
> > > consoleserver
> > > [19:57:02]     MeerK40t 0.8.4000 pkg console server on port: 23
> > > [19:57:02]     Could not start listening.
> > > ^Z
> > > [3]+  Stopped                 sudo meerk40t -c
> > > edcas@meerk40t:~ $

This is the problem you've been having all along!  Using Ctrl-Z like that only suspends a process, it does not kill it! So it is no wonder that you've been having trouble with getting a new MeerK40t instance to listen to the port -- you never actually quit the old one, and it continues to listen on the port!

What you really meant to do is Ctrl-C, that sends a keyboard interrupt to the process and signals that you as the user want to kill it. Only by quitting the process (or killing it) will the server release the port to be used again.

If you want to explore what Ctrl-Z is meant to help you do, then you should learn about the `fg` and `bg` commands. Somewhat useful job management in a Linux shell.

**Reply by edcasati** [reply]
The problem was that it crashed without me killing it in any way. My code was the culprit, I was opening and closing the connection with every sentence sent, and eventually it crashed the server on my Raspberry Pi. Stopping doing something dumb with my code fixed the problem.... open once at the beginning and leave it open until you close and leave.

**Reply by tiger12506** [reply]
Sure, no argument from me there. Just remember to use Ctrl-C for killing. Ctrl-Z just suspends a process in the shell so that you could (for example) run it in the background, it's really not the same thing.

**Comment by edcasati** [comment]
Is there a way to query the server for things like current position (absolute and relative, homed/not homed, current error status code, etc?

**Reply by tatarize** [reply]
I don't think so. There is `status` but that typically actually just queries the card for the status codes. 

Other than that there's a few different settings for the current parameters. You'd have current position, you wouldn't have relative/absolute since the position is always given in absolute even if you did some stuff to run it relative it stores the same internally. Homed would basically be the same as whether or not the absolute position is currently at `0,0` or not. And `status` would give you the current actual status code but the meaning of those codes differs from device to device.

In theory I think if you telnet into the system running grblcontrol the `?` command actually does some nice things with regard to the current status info.

I think it gives the position in X, Y. The currently set modals, and info like speed and power.

That information exists but there's was never really a clear need to query them, especially in a fairly easy to parse format.

There are `speed` and `power` commands. But, those are typically isolated.

Parameterwise there is:
```python
    "power",
    "raster_direction",
    "acceleration",
    "dot_length",
    "passes",
    "jog_distance",
    "raster_direction",
    "raster_preference_top",
    "raster_preference_right",
    "raster_preference_left",
    "raster_preference_bottom",
    "input_mask",
    "input_value",
    "output_mask",
    "output_value",
    "dpi",
    "raster_step_x",
    "raster_step_y",
    "speed",
    "rapid_speed",
    "dratio",
    "dwell_time",
    "frequency",
    "kerf",
    "passes_custom",
    "dot_length_custom",
    "acceleration_custom",
    "dratio_custom",
    "default",
    "dangerous",
    "output",
    "laser_enabled",
    "job_enabled",
    "ppi_enabled",
    "shift_enabled",
    "bidirectional",
    "advanced",
    "stopop",
    "overscan",
    "hatch_distance",
    "hatch_angle",
    "hatch_type",
    "penbox_value",
    "penbox_pass",
    "output_message",
    "input_message",
```
All of which could be set and checked. There's two different positions. These are `current` and `native` and they give you the actual position in whatever units you want within the field-coordinate-system so where the head currently in physical units. And `native` which gives the position of the head in device units which differs from device to device but is often needed to calculate relative moves etc.

So what do you want this status command to display, and how would you like to trigger it. `?` currently displays help info. `status` displays basically a query of the status code depending on different laser infos.

**Reply by edcasati** [reply]
It basically is 'bling' for the screen, not something that is really needed. Thanks for the extensive answer, I always learn a lot from you replies.

**Reply by tatarize** [reply]
https://github.com/meerk40t/meerk40t/pull/1978 

I'm more than happy to add whatever you'd need for your workflow there. In theory if you ran it as a plugin you could access that data as `context.device.current` or `context.device.native` and I've no qualms adding that stuff. I assume easily parsible is probably better. So what bits of info do you need out of that stuff.

```
[15:36:54] devinfo
[15:36:54]     0.0,0.0;0,0;
[15:37:02] tool relocate
[15:37:02] move_absolute 109.99999999999999mm 89.99999999999999mm
[15:37:07] devinfo
[15:37:07]     283832.08499993343,232190.50499994555;4331,3543;
```

That's current position (in meerk40t coords) and current position in native coords. With a `;` between the two and some `,` between the coords.

Is the current PR 1978 to add in a quick status command. I assume it's probably best to flag each thing with what it means. Sort of like the grbl `?` command:

<Idle|MPos:83.276,-57.876,0.000|FS:0,20|Pn:Z|WCO:0.000,-65.000,0.000>okok
<Run|MPos:83.272,-57.872,0.000|FS:100,0|Pn:Z|Ov:100,100,100>ok
<Run|MPos:83.260,-57.864,0.000|FS:100,0|Pn:Z>ok
<Run|MPos:83.244,-57.852,0.000|FS:100,0|Pn:Z>ok
<Run|MPos:83.232,-57.844,0.000|FS:100,0|Pn:Z>ok

Which includes a couple things like machine state, and FS which is speed and power, and some modals.

**Comment by edcasati** [comment]
Do you already provide the capability to turn the water on/off, and the fan on/off?
My little app runs on a raspberry pi, so I can use the GPIO ports connected to relays if necessary.

Going slow but well. I can only work on this occasionally because my body/brain are not working so good anymore. (I have Parkinsons) Like I said earlier, this exercises my brain which it needs badly... I don't want to sit and just watch TV all day! Thanks for the support.

**Reply by tiger12506** [reply]
The m2nano doesn't really support any of that, so it tends to be outside the scope of MeerK40t itself. i.e. you have an external python script already, that is calling into meerk40t -- it wouldn't make much sense to have that external script call into meerk40t just for meerk40t to call back out into another external script.
That said there are numerous ways of controlling GPIO on raspberry pis and there is a pretty strong plugin framework for MeerK40t if you need some sort of information from meerk40t in order to make that happen. If I were you, though, I'd start by figuring out how to control GPIO on the pi first, without involving meerk40t, and then use that to control relays as you suggested.

**Reply by tatarize** [reply]
There's not really many controller boards that allow this. In theory, you could hook this equally to the GPIO a grbl. The ruida controller has a flag for `blower` which lets you turn the air assist on and off. There were too many different ways to turn these things on and off. To strictly hook them directly into meerk40t. The current hope is to perhaps send a signal that is intended to turn the air on/off, but without anything to accept this signal there wasn't much work done.

Somebody did ask for a way to do this with the M2 through packet sniffing and weirdly you can send a packet to the backend that will not process on the M2 in-case you read packets in-line and want to use them to activate something like air, but it's an extremely specialized idea there. Equally there's stuff like water temperature sensors that could be added but would require an addon to work successfully since the ability to interact with the rPI GPIO and the ability to read temperature from some otherwise unknown sensor isn't something that should be built into the core of the program.

You'd really want to be able to turn on the air and water from various relays. This can be hooked to the M2 but it just turns on when you start cutting and stays on for like 8 seconds after you're done (and needs a relay since you can't actually use much of that current). But, there's no command to trigger that stuff.

Usually this stuff is just done manually because you *always* want your water running and the only thing there would be to maybe check a flowometer and kill the laser if the flow somehow stopped. And there's a pretty limited need to turn air on and off, there's some need to turn it off for some acrylic cuts but you'd usually just leave it off for the whole thing, I don't know a use case that would want it on for some cuts but off for others.

**Reply by edcasati** [reply]
The water and air switches are insurance. The job won't start unless you have at least checked them off the list by pressing their respective switches, even if they don't do anything. I need all the reminders I  can manage.

**Reply by tatarize** [reply]
If you're only hooked in with console manipulations you might not be hooked into the code deep enough to really access what those would need to be in the end, which is signals probably `air_assist` and `water_pump` but there's still absolutely nothing that cares about those signals or anything that issues them. There are commands in GRBL for M7, M8, and M9 which are `coolant control` and these will send a signal called `coolant` but only within the job. There are realtime commands for `0xA0` which toggles the flood coolant and `0xA1` which toggles the mist coolant but these currently do nothing.

As for monitoring things there's something to the idea of water temperature. Automatically allowing a job to pause if the water exceeds a threshold but there's so many different thermometers that it would need to also be an addon.

Your current methodology is just to TCP into the console (by the way, the latest version of `main` branch has a `devinfo` command that gives you the device position in the form real position in 1/65535 of an inch, and native position (device dependent)), just won't have access to a `context` and thus can't register anything or read signals etc.

Again, see: https://github.com/meerk40t/meerk40t-example-extension for how to hook into meerk40t as a first order citizen.

**Comment by edcasati** [comment]
I'm still working on the 5 inch screen interface.
I originally wrote it in Tkinter, but I didn't really like the appearance, and the page switching was really slow.
So I rewrote the whole thing in Kivy. The page switching is great, but there is a problem somewhere that I haven't been able to resolve.

https://www.youtube.com/watch?v=wDjwb59qVQA

**Reply by tiger12506** [reply]
Probably would need to see the code for the Kivy version to give you some hints in the right direction.

**Reply by edcasati** [reply]
After making what little hair I have left on my head even more gray....

Fixed it.

If you have a touchscreen that records touches in more than one place at a time, this is the fix that worked for me.:

in ~/.kivy/config.ini

replace whatever is in the [input] section with:

[input]
mouse = mouse
wm_touch = wm_touch
wm_pen = wm_pen
